API: espi: Add API for Enhanced Serial Peripheral Interface

albertofloyd opened the pull request:
This API defines following calls for eSPI drivers
For more information regarding this refer to spec  https://www.intel.com/content/dam/support/us/en/documents/software/chipset-software/327432-004_espi_base_specification_rev1.0_cb.pdf

- espi_set_config
- espi_get_channel_status
- espi_read_request
- espi_write_request
- espi_send_vwire
- espi_receive_vwire
- espi_send_oob
- espi_receive_oob
- espi_flash_read
- espi_flash_write
- espi_flash_erase

There is a sample eSPI driver for MEC1501 family that support eSPI bus operation for eSPI channel 1.
The driver doesn't support low level operations for eSPI channel 0 since they are handled by HW.
Channel 2 & Channel 3 can be added in the future. 

A sample driver eSPI test that showcase the API and callback mechanism

Signed-off-by: Jose Alberto Meza <jose.a.meza.arellano@intel.com>

albertofloyd committed:
drivers: espi: Add Kconfig for eSPI driver

Adds the Kconfig for generic eSPI drivers

Signed-off-by: Jose Alberto Meza <jose.a.meza.arellano@intel.com>

andrewboie commented on include/espi.h:
```diff
@@ -0,0 +1,446 @@
+	espi_api_get_pending_int get_pending_int;
+};
+
+__syscall int espi_config(struct device *dev, struct espi_cfg *cfg);
```
Need documentation comments for all APIs

dcpleung replied:
Also, are all these APIs supposed to be `INTERNAL_HIDDEN`? I would assume that applications are using these.

albertofloyd replied:
Fixed

dcpleung commented on include/espi.h:
```diff
@@ -0,0 +1,446 @@
+/*
+ * eSPI VW channel packets contain 4 VWire states.
+ */
+#define ESPI_VW_SRC_ID0         0
```
Since you are using `VWIRE_CHANNEL` below, it is better to do `ESPI_VWIRE_SRC_ID0` to be consistent.

dcpleung commented on include/espi.h:
```diff
@@ -0,0 +1,446 @@
+enum espi_bus_event {
+	ESPI_RESET = 1,
+	CHANNEL_READY,
+	VW_RECEIVED,
```
`VWIRE_RECEIVED`

dcpleung commented on include/espi.h:
```diff
@@ -0,0 +1,446 @@
+	OOB_RECEIVED,
+};
+
+enum espi_vw_signal {
```
`espi_vwire_signal`

dcpleung reviewed (changes requested):
Some minor things to be consistent throughout the file.

tbursztyka commented on include/espi.h:
```diff
@@ -0,0 +1,446 @@
+	FLASH_CHANNEL		= BIT(3),
+};
+
+enum espi_peripheral {
```
I don't think you need to expose this

albertofloyd replied:
Agree this could part of eSPI driver KConfig

tbursztyka commented on include/espi.h:
```diff
@@ -0,0 +1,446 @@
+ * @brief ESPI I/O mode capabilities
+ */
+enum espi_io_mode {
+	DUAL_LINES = BIT(0),
```
Add ESPI_IO_MODE_ as prefix on all values here.

tbursztyka commented on include/espi.h:
```diff
@@ -0,0 +1,446 @@
+/**
+ * @brief ESPI frequency capabilities
+ */
+enum espi_freq {
```
I would stick with a more generic attribute in espi_config instead of an enum. (The specs provide register reference, but that's for the hardware side, no need to expose its current limitation to the public API).

tbursztyka commented on include/espi.h:
```diff
@@ -0,0 +1,446 @@
+ */
+
+enum espi_channel {
+	PERIPHERAL_CHANNEL	= BIT(0),
```
Add ESPI_CHANEL_ as prefix to all values here (remove _CHANNEL from existing names obviously)

tbursztyka commented on include/espi.h:
```diff
@@ -0,0 +1,446 @@
+	OOB_RECEIVED,
+};
+
+enum espi_vw_signal {
```
Does this need to be exposed publicly? I mean, besides the master and slave driver, the user won't have to manipulate these afaik.

Anyway, add ESPI_VWIRE_SIGNAL_ as prefix to all values.

albertofloyd replied:
Most of eSPI traffic including eSPI vwire signals are can be handled within the eSPI drivers but certain events are required to be exposed to user apps similar to GPIO control or events that trigger eSPI-unrelated flows such as power management, thermal management, reset sequence

+-------+                  +-------------+   +------+     +---------+
|  App  |                  | eSPI driver |   |  HW  |     |eSPI Host|
+---+---+                  +-------+-----+   +---+--+     +----+----+
    |                              |             |             |
    |   espi_init_callback         |             |             |
    +----------------------------> |             |             |
    |   espi_add_callback          |             |
    +----------------------------->+             |
    |                              |             |  VW CH ready|  eSPI host
    |                              |    IRQ      +<------------+  enables VW
    |                              | <-----------+             |  channel
    |                              |             |             |
    |                              | Processed   |             |
    |                              |             |             |
    |                              |             |             |
    |                              |             |             |
    |                              |             |             |
    |                              |             |             |
    |                              |             |             |
    |                              |             | HOST_RST    |  eSPI host
    |                              |             <-------------+  send VWire
    |                              +<------------+             |
    +<-----------------------------+ callback    |             |
    | App reset host-related       |             |             |
    | data structures              |             |             |
    |                              |             |             |
    |                              |             |   C10       |
    |                              |             +<------------+
    |                              <-------------+             |
    <------------------------------+             |             |
    | App executes                 |             |             |
    + power policy                 +             +             +

tbursztyka commented on include/espi.h:
```diff
@@ -0,0 +1,446 @@
+}
+
+__syscall int espi_enable_callback(struct device *dev,
+				   struct espi_event *cfg);
```
s/cfg/event I guess

tbursztyka commented on include/espi.h:
```diff
@@ -0,0 +1,446 @@
+					struct espi_callback *callback,
+					bool set);
+typedef int (*espi_api_enable_callback)(struct device *dev,
+					struct espi_event *cfg);
```
s/cfg/event ?

tbursztyka commented on include/espi.h:
```diff
@@ -0,0 +1,446 @@
+typedef int (*espi_api_enable_callback)(struct device *dev,
+					struct espi_event *cfg);
+typedef int (*espi_api_disable_callback)(struct device *dev,
+					struct espi_event *cfg);
```
same

tbursztyka reviewed (commented):
Hard to tell at the moment where is the end user API (as there is nothing to make any transaction so far). Looking forward to seeing this added.

Looks like many functions are mostly meant to be used between the controller driver and the slave drivers (get/set_vwire, channel_status etc...). As already pin-pointed you'll need to document all that.

Best also it to follow with a test application (it does not have to build yet): to show how the API should be used on the end-user side, with comments explaining steps etc... It will help you to expose the use-cases the API will address as well as it will help us understanding these and perhaps proposing other ones.

I saw you took some inspiration on gpio callbacks. Note these were designed like that due to some controller existing on memory constrained devices. I bet eSPI is not going to land on such device, so if you feel such callback approach is too limited in performance for you, don't hesitate to come with something that would suit eSPI better.

albertofloyd committed:
API: espi: Add API for Enhanced Serial Peripheral Interface

    This API defines following call for eSPI bus drivers

    - espi_set_config
    - espi_get_channel_status
    - espi_send_read_request
    - espi_send_write_request
    - espi_send_vwire
    - espi_receive_vwire
    - espi_send_oob
    - espi_receive_oob
    - espi_flash_read
    - espi_flash_write
    - espi_flash_erase

Signed-off-by: Jose Alberto Meza <jose.a.meza.arellano@intel.com>

albertofloyd commented on include/espi.h:
```diff
@@ -0,0 +1,557 @@
+ * @retval 0 If successful.
+ * @retval -EIO General input / output error, failed to configure device.
+ * @retval -EINVALID invalid capabilities, failed to configure device.
+ * @retval -ENOTSUP capability not supported by eSPI slave
```
Intention was to have eSPI API to be implemented for eSPI slaves. But I think in can be used in both case eSPI master & slaves. I have changed the documentation and indicating few differences in espi_config

albertofloyd commented on include/espi.h:
```diff
@@ -0,0 +1,557 @@
+/**
+ * @internal
+ */
+static inline int z_impl_espi_get_pending_int(struct device *dev)
```
Removed

albertofloyd committed:
drivers: espi: Add driver for microchip XEC family

Add eSPI Microchip XEC driver
Include support for peripheral & virtual wires (channel 0-1)
OOB and flash support can be added in the future

Fix compilation error in pinmux driver

Signed-off-by: Jose Alberto Meza <jose.a.meza.arellano@intel.com>

albertofloyd committed:
dts: espi: Add dts bindings for eSPI

Add generic device tree bindings for eSPI devices

Signed-off-by: Jose Alberto Meza <jose.a.meza.arellano@intel.com>

albertofloyd committed:
dts: espi: mchp: Add dts bindings for XEC eSPI driver

Add device for eSPI driver for MEC family

Signed-off-by: Jose Alberto Meza <jose.a.meza.arellano@intel.com>

albertofloyd committed:
ext: hal: mchp: Fix compilation for MEC1501 HAL macros

Correct HAL macros related to eSPI block
Add GIRQ bit definitions for VWires and some peripherals

Signed-off-by: Jose Alberto Meza <jose.a.meza.arellano@intel.com>

albertofloyd commented on include/espi.h:
```diff
@@ -0,0 +1,557 @@
+ *
+ * @retval 0 If successful.
+ * @retval -EIO General input / output error, failed to configure device.
+ * @retval -EINVALID invalid capabilities, failed to configure device.
```
Fixed

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,552 @@
+
```
Remove blank line

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,552 @@
+ * @{
+ */
+
+
```
remove blank line

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,552 @@
+	ESPI_VWIRE_SIGNAL_SLV_BOOT_DONE,
+	ESPI_VWIRE_SIGNAL_HOST_RST_ACK,
+	ESPI_VWIRE_SIGNAL_RST_CPU_INIT,
+	ESPI_VWIRE_SIGNAL_SMI, /* System management interrupt */
```
Don't put comments on the right. Put them on top like you did on line 179

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,552 @@
+ * @brief ESPI event
+ */
+struct espi_event {
+	/** Event type. */
```
remove the dot after type

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,552 @@
+					struct espi_callback *cb,
+					struct espi_event espi_evt);
+
+
```
remove one blank line

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,552 @@
+ * Note: To help setting it, see espi_init_callback() below
+ */
+struct espi_callback {
+	/** This is meant to be used in the driver only
```
Finish the comment in the same line

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,552 @@
+
+/**
+ * @typedef espi_callback_handler_t
+ * @brief Define the application callback handler function signature
```
Finish all the sentences in "@brief" and "@params" with a dot.

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,552 @@
+
+
+/**
+ * @brief Configure operation of a eSPI controller
```
Finish sentences with a dot

albertofloyd replied:
Finished sentences with more than a line with a dot.

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,552 @@
+					u8_t *level);
+
+static inline int z_impl_espi_get_vwire(struct device *dev,
+				enum espi_vwire_signal signal, u8_t *level)
```
indent more +1

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,552 @@
+}
+
+/**
+ * Callback usage model
```
Callback model

franciscomunoz reviewed (commented):
fix the the cosmetic issues

albertofloyd commented on include/espi.h:
```diff
@@ -0,0 +1,557 @@
+}
+
+/**
+ * @brief Query to see if it a channel is ready
```
Not, there use cases where APPs need to check channel availability for non-eSPI traffic, mainly related to boot handshake.  i.e. eSPI master/slave needs to confirm a channel has been disabled or enabled

tbursztyka commented on include/espi.h:
```diff
@@ -0,0 +1,550 @@
+typedef int (*espi_api_manage_callback)(struct device *dev,
+				struct espi_callback *callback,
+				bool set);
+typedef u32_t (*espi_api_get_pending_int)(struct device *dev);
```
that can be removed then.

tbursztyka reviewed (changes requested):
Just blocking it in order to avoid the PR being merged too early without any commit for a driver implementing the API, a test and eventually a sample showing its usage from a user perspective.

Maybe it's just me, but I don't understand how transaction are supposed to happen with the current proposal. So far, here it configures, set/get signals between master and slave but that's all. I don't see how specs chapter 5 fits within this PR for instance.

albertofloyd commented:
eSPI controllers usually do implement the transaction layer, so section 5.1 & 5.3-5.4 won't be reflected in the driver API, instead only section 5.2 should be exposed. 

Section 5.2.2. is covered by current API set/get  "eSPI vwire signal" which are actually performing a packet transfer in eSPI bus as shown in Figure 42 over logical channel 1.  I could rename espi_set_vwire to espi_send_vwire_packet to clear the semantics than an actual transfer is happening over the bus.  This just a way we have implemented in our project and how chromium is doing it.
https://chromium.googlesource.com/chromiumos/platform/ec/+/master/include/espi.h

Note there are intentionally some missing APIs to support other kind of packets, I can add these even though they are not widely used. 
spec section 5.2.3  Logical Channel 2 API
espi_send_OOB(u8_t *buf, u16_t len, UINT16 timeout);
espi_receive_OOB(u8_t *buf, u16_t len, UINT16 timeout);

spec section 5.2.4   Logical channel 3 APIs
espi_flash_read(u8_t* buf, u32_t flash_addr, u16_t len, u16_t timeout);
espi_flash_write(u8_t* buf, u32_t flash_addr, u16_t len, u16_t timeout);
espi_flash_erase(u8_t* buf, u32_t flash_addr, u16_t len, u16_t timeout);

Alternatively will be to add a eSPI_flash driver implementing flash driver API and eSPI_I2C driver  implementing I2C driver API

albertofloyd committed:
samples: drivers: espi: Add eSPI driver sample app

Show eSPI API driver usage, including early configuration,
send of virtual wire packets and callback mechanism

Signed-off-by: Jose Alberto Meza <jose.a.meza.arellano@intel.com>

albertofloyd commented:
@tbursztyka , @andrewboie  please re-review this PR. 
As requested I have updated eSPI header with additional operations, sample eSPI driver for MEC has been  added & eSPI driver sample test app.

franciscomunoz commented on ext/hal/microchip/mec/mec1501/component/espi_vw.h:
```diff
@@ -420,7 +420,7 @@ inline void
 			  enum espi_msvw_irq_sel isel)
 {
-	volatile uint8_t *psrc = (volatile uint8_t *)p->SRC_IRQ_SEL;
+	volatile uint8_t *psrc = (volatile uint8_t *)&p->SRC_IRQ_SEL;
```
Is this change required?

albertofloyd replied:
Yes, this is needed to properly enabled interrupt though it's convoluted
@scottwcpg can we clean this up this way  
       p->SRC_IRQ_SEL = (isel << src);

franciscomunoz commented on ext/hal/microchip/mec/mec1501/component/espi_vw.h:
```diff
@@ -440,7 +440,7 @@ static __attribute__ ((always_inline))
 mec_espi_msvw_set(ESPI_MSVW_REG * p, enum espi_msvw_src src, uint8_t src_val)
 {
-	volatile uint8_t *psrc = (volatile uint8_t *)p->SRC;
+	volatile uint8_t *psrc = (volatile uint8_t *)&p->SRC;
```
i don't think this change is needed ?

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,769 @@
+ * Used to register a callback in the driver instance callback list.
+ * As many callbacks as needed can be added as long as each of them
+ * are unique pointers of struct espi_callback.
+ * Beware such structure should not be allocated on stack.
```
I think this line is wrong.

tbursztyka replied:
Sounds fine to me. Callback structures have to be statically allocated (as for gpio callbacks, etc...) It's worth noticing it in the documentation.

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,769 @@
+ * If this eSPI controller is acting as master, the values set here
+ * will be used by eSPI master to determine minimum common capabilities with
+ * eSPI slave then send via SET_CONFIGURATION command
+ *
```
Please indicate in the drawing the flows for both scenarios.  Maybe add a row indicating the Roles of the eSPI controller.

albertofloyd replied:
Flows already were already indicated, added clarifications and indication when eSPI driver programs the corresponding eSPI controller

franciscomunoz reviewed (commented):
Minor observations: There are a few @params explanations which do not end with a dot.

franciscomunoz commented on drivers/espi/Kconfig:
```diff
@@ -0,0 +1,98 @@
+		ESPI Controller supports virtual wires channel.
+
+config ESPI_CHANNEL_OOB
+	bool "IRQ Priority for ESPI Controller"
```
ESPI oob channel

franciscomunoz commented on drivers/espi/Kconfig:
```diff
@@ -0,0 +1,98 @@
+	int "IRQ Priority for ESPI Controller"
+	default 3
+	help
+		IRQ Priority for ESPI Controller.
```
remove dot

franciscomunoz commented on drivers/espi/Kconfig:
```diff
@@ -0,0 +1,98 @@
+	bool "ESPI peripheral channel"
+	default y
+	help
+		ESPI Controller supports peripheral channel.
```
remove dot

franciscomunoz commented on drivers/espi/Kconfig:
```diff
@@ -0,0 +1,98 @@
+	bool "ESPI flash channel"
+	default y
+	help
+		ESPI Controller supports virtual wires channel.
```
remove dot

franciscomunoz commented on drivers/espi/Kconfig:
```diff
@@ -0,0 +1,98 @@
+	bool "IRQ Priority for ESPI Controller"
+	default n
+	help
+		ESPI Controller supports OOB channel.
```
remove dot

franciscomunoz commented on drivers/espi/Kconfig:
```diff
@@ -0,0 +1,98 @@
+	bool "ESPI flash channel"
+	default n
+	help
+		ESPI Controller supports flash channel.
```
remove dot

franciscomunoz commented on drivers/espi/Kconfig:
```diff
@@ -0,0 +1,98 @@
+# Kconfig - DMA configuration options
+#
```
remove

franciscomunoz commented on drivers/espi/Kconfig:
```diff
@@ -0,0 +1,98 @@
+# Kconfig - DMA configuration options
+#
+#
```
remove

tbursztyka replied:
It's does not need 2 #, but 1 is nice. comments don't have to be coalesced one after each other.

franciscomunoz commented on dts/bindings/espi/microchip,xec-espi.yaml:
```diff
@@ -0,0 +1,45 @@
+
+    agg_io_irq:
+      type: int
+      description: soc nvic mapping
```
Aggregated isr for eSPI block

franciscomunoz commented on drivers/espi/Kconfig.xec:
```diff
@@ -0,0 +1,19 @@
+	bool "MEC/CEC Microchip ESPI driver"
+	depends on ESPI
+	help
+	Enable the Microchip MEC/MEC ESPI driver.
```
Change the word MEC/CEC to XEC in all occurrences.  This is because the current CEC does not have an eSPI block

franciscomunoz commented on drivers/espi/Kconfig.xec:
```diff
@@ -0,0 +1,19 @@
+# Kconfig.xec - Microchip CEC/MEC ESPI configuration options
```
CEC/MEC to XEC

franciscomunoz replied:
The CECs don't have the ESPI block. That's why it is better to say XEC in my opinion like we have been doing it.

franciscomunoz commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+	}
+
+	/* Set IO mode */
+	iomode = (iomode >> 1);
```
does not make sense

franciscomunoz commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+	 * de-assertion and after pinmux
+	 */
+	ESPI_EIO_BAR_REGS->IO_ACTV = 1;
+	printk("eSPI block actived succesfully\n");
```
Remove this printk

tbursztyka replied:
Or switch to LOG_DBG()

franciscomunoz commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+
+struct espi_isr {
+	u32_t girq_bit;
+	void (*the_isr)(struct device *dev);
```
it is up to you, but I would name it "isr" instead.

tbursztyka replied:
+1

franciscomunoz commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+		(struct espi_xec_data *) (dev->driver_data);
+	struct espi_event evt = { ESPI_BUS_PERIPHERAL_NOTIFICATION, 0, 0 };
+
+	evt.evt_details = 0;
```
Use a macro instead of magic numbers

franciscomunoz commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+		(struct espi_xec_data *) (dev->driver_data);
+	struct espi_event evt = { ESPI_BUS_PERIPHERAL_NOTIFICATION, 0, 0 };
+
+	evt.evt_details = 1;
```
Use a macro instead of magic numbers

franciscomunoz replied:
In addition, add the UART2 isr.

franciscomunoz commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+		(struct espi_xec_data *) (dev->driver_data);
+	struct espi_event evt = { ESPI_BUS_PERIPHERAL_NOTIFICATION, 0, 0 };
+
+	evt.evt_details = 2;
```
Use a macro instead of magic numbers

franciscomunoz commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+		(struct espi_xec_data *) (dev->driver_data);
+	struct espi_event evt = { ESPI_BUS_PERIPHERAL_NOTIFICATION, 0, 0 };
+
+	evt.evt_details = 3;
```
Use a macro instead of magic numbers

franciscomunoz commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+	.bus_girq_id = MCHP_ESPI_IO_GIRQ,
+	.vw_girq_id = MEC_ESPI_MSVW_00_06_GIRQ,
+	.pc_girq_id = MCHP_KBC_GIRQ,
+};
```
Add space

franciscomunoz commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+	MCHP_GIRQ_ENSET(config->vw_girq_id) = BIT(0) | BIT(1) | BIT(2) |
+		BIT(5) | BIT(8) | BIT(12);
+
+	MCHP_GIRQ_ENSET(config->pc_girq_id) = BIT(0) | BIT(1);
```
If this is GIRQ15 it means you are enabling UART0 and UART1. Remember we moved to uart2 which should be BIT(4). Also add a comment indicating this is a different block

tbursztyka replied:
all these BIT() call should be hidden behind macros

albertofloyd replied:
These are virtual peripherals exposed by the eSPI endpoint, there is no direct relation to SoC ones. These ISRs will trigger when eSPI master attempts to access the respective peripheral in the eSPI endpoint and the eSPI controller HW detects that traffic in the bus.

franciscomunoz commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+#ifdef CONFIG_ESPI_CHANNEL_OOB
+	ESPI_CAP_REGS->GLB_CAP0 |= MCHP_ESPI_GBL_CAP0_OOB_SUPP;
+	ESPI_CAP_REGS->OOB_CAP |= MCHP_ESPI_OOB_CAP_MAX_PLD_SZ_73;
+	ESPI_OOB_REGS->TX_ADDR_MSW = 0x1FFFul;
```
I'm not sure if this should stay in hex or behind a mask. What would be the implication if this is changed to a Macro?

tbursztyka replied:
better as a macro indeed, same for the numbers below

franciscomunoz commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+
+const struct espi_isr peripherals_isr[] = {
+	{ BIT(0), uart0_isr },
+	{ BIT(1), uart1_isr },
```
add uart2 isr

tbursztyka commented on include/espi.h:
```diff
@@ -0,0 +1,769 @@
+	const struct espi_driver_api *api =
+		(const struct espi_driver_api *)dev->driver_api;
+
+	return api->flash_read(dev, pckt);
```
For better consistency, maybe you want to test flash_read here (and return -ENOTSUP accordingly then). As you do it for manage_callback etc...

(either way in fact: either you force each and every driver to implement the functions - and return -ENOTSUP when relevant - or you allow them to let some api function being NULL).


Same on read/write request actually.

albertofloyd replied:
@tbursztyka  Verified this is addressed in the latest

tbursztyka commented on drivers/espi/Kconfig:
```diff
@@ -0,0 +1,98 @@
+	bool "ESPI slave driver"
+	default y
+	help
+		Enables eSPI driver in slave mode
```
intend on help message is actually one tab + 2 spaces. Apply that everywhere

tbursztyka commented on drivers/espi/Kconfig:
```diff
@@ -0,0 +1,98 @@
+		ESPI Controller supports peripheral channel.
+
+config ESPI_CHANNEL_VWIRE
+	bool "ESPI flash channel"
```
the option name and the title don't match it seems ?

tbursztyka commented on drivers/espi/Kconfig:
```diff
@@ -0,0 +1,98 @@
+		ESPI Controller supports flash channel.
+
+config ESPI_PERIPHERAL_UART
+	bool "UART"
```
"UART peripheral"

Add "peripheral" to all relevant option titles below as well.

tbursztyka commented on dts/bindings/espi/espi.yaml:
```diff
@@ -0,0 +1,22 @@
+---
```
Add intel copyright

tbursztyka commented on dts/bindings/espi/espi.yaml:
```diff
@@ -0,0 +1,22 @@
+      category: required
+      description: compatible strings
+      generation: define
+    label:
```
add an empty line between each properties, so here before 'label'

tbursztyka commented on dts/arm/microchip/mec1501hsz.dtsi:
```diff
@@ -132,6 +132,11 @@
 			clock-frequency = <I2C_BITRATE_STANDARD>;
```
rewrite a bit your commit title: "dts: espi: Add bindings for microchip XEC driver"

tbursztyka commented on drivers/espi/CMakeLists.txt:
```diff
@@ -2,3 +2,5 @@
 
```
Commit title: "drivers: espi: Add driver for microchip XEC family"

albertofloyd replied:
These are the generic definitions for all eSPI drivers. I think you meant the commit after this

tbursztyka commented on drivers/espi/CMakeLists.txt:
```diff
@@ -2,3 +2,5 @@
 
 zephyr_library()
 
+zephyr_library_sources_ifdef(CONFIG_ESPI_XEC		espi_mchp_xec.c)
```
Having the name of the manufacturer in the driver filename is not mandatory. But if you want it, you will have to write it all like "espi_microchip_xec.c"

But espi_xec.c would be totally fine as well

albertofloyd replied:
Agree, but we already agreed this as file naming for Microchip drivers

tbursztyka replied:
ok missed that. fine by me

tbursztyka commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+#include <espi.h>
+#include "espi_utils.h"
+
+#define M2S             1ul
```
Always prefix your macros with something that identify the owner so here ESPI_XEC_ for instance

tbursztyka commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+	u8_t dir;
+};
+
+enum MCHP_MSVW_REGS {
```
enum main name are in small letters (and same as macros: prefix them)

tbursztyka replied:
this comment has not be applied.

albertofloyd replied:
@tbursztyka  addressed in most recent submission

tbursztyka commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+
+	/* Set frequency */
+	cap1 &= ~MCHP_ESPI_GBL_CAP1_MAX_FREQ_MASK;
+	switch (cfg->max_freq) {
```
have an empty line before the switch. it has to be coalesced if only the previous assignement is being used in the if ()/for ()/while ()/ switch ()

tbursztyka replied:
same here

albertofloyd replied:
Addressed

tbursztyka commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+			MCHP_ESPI_GBL_CAP1_IO_MODE_POS;
+		cap1 |= (iomode << MCHP_ESPI_GBL_CAP1_IO_MODE_POS);
+	}
+	ESPI_CAP_REGS->GLB_CAP1 = cap1;
```
Add an empty line after a '}', unless it's another '}' etc... apply that everywhere.

albertofloyd replied:
Adressed

tbursztyka commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+}
+
+static bool espi_xec_channel_ready(struct device *dev,
+	enum espi_channel ch)
```
indentation should be below last '(' + 1 space. Same for function calls (as long as it fits the 80 chars). Apply that everywhere.

tbursztyka commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+static int espi_xec_read_request(struct device *dev,
+					struct espi_request_packet req)
+{
+	return -ENOTSUP;
```
See my comment in espi.h : either you allow the driver to set the api function pointer to NULL or you do as you did here. but not both.

tbursztyka commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+	}
+	if (signal_info.dir == M2S) {
+		ESPI_MSVW_REG *reg = &(ESPI_M2S_VW_REGS->MSVW00) + xec_id;
+		*level = ((reg->SRC >> (src_id << 3)) & 0x01ul);
```
it's better to have a self documented macro name instead of some "magic numbers" like here.

tbursztyka commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+	ESPI_EIO_BAR_REGS->EC_BAR_MBOX = 0x03600001;
+	ESPI_EIO_BAR_REGS->EC_BAR_PORT92 = 0x00920001;
+	ESPI_EIO_BAR_REGS->EC_BAR_UART_0 = 0x03F80001;
+	ESPI_EIO_BAR_REGS->EC_BAR_P80CAP_0 = 0x00810001;
```
All there BARs should come from DTS only.

Btw, you have ESPI controller device driver (and dts bindings), do you need any espi device then? (in dts you would have a espi device with an espi controller parent then).

tbursztyka commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+static struct espi_xec_data espi_xec_data;
+static const struct espi_xec_config espi_xec_config = {
+	.base_addr = DT_MICROCHIP_XEC_ESPI_0_BASE_ADDRESS,
+	.bus_girq_id = MCHP_ESPI_IO_GIRQ,
```
Same as BARs, IRQs should all come from DTS

albertofloyd replied:
Fixed. Defined all group IRQs in DTS

tbursztyka commented on ext/hal/microchip/mec/mec1501/MEC1501hsz.h:
```diff
@@ -569,9 +569,9 @@ typedef enum IRQn {
 
 /* eSPI Virtual Wire registers in IO component */
-#define ESPI_IO_VW_REGS      ((ESPI_IO_VW_Type *) ESPI_IO_VW_BASE)
+#define ESPI_IO_VW_REGS      ((ESPI_IO_VW_Type *) (ESPI_IO_VW_BASE))
```
Are these change coming from an update of the microchip ext/hal or are these manual edition? 

If it's the later, an future update of the ext/hal will erase these.

I don't have a solution on how to act here though.

tbursztyka commented on drivers/pinmux/pinmux_mchp_xec.c:
```diff
@@ -135,6 +135,7 @@ static const struct pinmux_xec_config pinmux_xec_port040_076_config = {
 };
 
 DEVICE_AND_API_INIT(pinmux_xec_port040_076, CONFIG_PINMUX_XEC_GPIO040_076_NAME,
+		    &pinmux_xec_init,
```
merge it with the commit introducing the driver

albertofloyd replied:
Addressed

tbursztyka commented on samples/drivers/espi/src/main.c:
```diff
@@ -0,0 +1,339 @@
+#define SLPS3_DLY               30000
+
+/* Board specific definitions */
+#ifdef CONFIG_ESPI_XEC
```
I guess it would be better to just have a generic macro name. See what's been done for samples/driver/watchdog for instance. The generic name comes from soc's dts_fixup.h

tbursztyka commented on samples/drivers/espi/src/main.c:
```diff
@@ -0,0 +1,339 @@
+/*
+ * Copyright (c) 2012-2014 Wind River Systems, Inc.
```
wrong ©

tbursztyka commented on samples/drivers/espi/src/main.c:
```diff
@@ -0,0 +1,339 @@
+
+#ifdef CONFIG_GPIO_XEC
+	printk("eSPI test - I/O initialization...");
+	GPIO_CTRL_REGS->CTRL_0012 = MCHP_GPIO_CTRL_MUX_F0;
```
That should not happen here. Maybe on the soc or board source level then?

tbursztyka commented on samples/drivers/espi/src/main.c:
```diff
@@ -0,0 +1,339 @@
+	GPIO_CTRL_REGS->CTRL_0115 = MCHP_GPIO_CTRL_MUX_F0;
+#endif
+	ret = gpio_pin_configure(gpio_devA, BOARD_PWRGD_PIN, GPIO_DIR_IN);
+	if (ret)
```
always have '{ ... }' even for one liner statement in if/for/while. Apply that everywhere.

tbursztyka commented on samples/drivers/espi/src/main.c:
```diff
@@ -0,0 +1,339 @@
+	if (gpio_devB) {
+		printk("%s FOUND!\n", PORTB_LABEL);
+	}
+	gpio_devC = device_get_binding(PORTC_LABEL);
```
All these gpios are most probably necessary for your board, but will not be for another board (maybe less or maybe more will be required). You'll have to find a way to generalize this

albertofloyd replied:
Put all board-specific in a Kconfig and board-specific prj.conf which showscase  eSPI minimal API functionality

tbursztyka reviewed (changes requested):
trivial style issues 

Some concerns about DTS integration

albertofloyd commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+	ESPI_EIO_BAR_REGS->EC_BAR_KBC |= MCHP_ESPI_IO_BAR_HOST_VALID;
+	ESPI_EIO_BAR_REGS->EC_BAR_ACPI_EC_0 |= MCHP_ESPI_IO_BAR_HOST_VALID;
+	ESPI_EIO_BAR_REGS->EC_BAR_ACPI_EC_1 = 0x06A00001;
+	ESPI_EIO_BAR_REGS->EC_BAR_MBOX = 0x03600001;
```
Removed some of the peripherals not intended to be exposed in the driver. Added logical bars address definitions

dbkinder commented on samples/drivers/espi/README.rst:
```diff
@@ -0,0 +1,45 @@
+
+This sample demonstrates how to use the Enhanced Serial Peripheral Interface
+(eSPI) API.
+Indicating how to configure/select controller capabilities as part of simplest
```
Write as a complete sentence:
```
It shows how to configure and select controller capabilities as part of a simple
eSPI handshake that includes exchanging virtual wire packets.
```

dbkinder commented on samples/drivers/espi/README.rst:
```diff
@@ -0,0 +1,45 @@
+Indicating how to configure/select controller capabilities as part of simplest
+eSPI handshake that includes exchange of virtual wire packets.
+
+Standard platform signals are sent virtual wire packets over the bus and
```
How about:
```
Standard platform signals are sent as virtual wire packets over the bus.
Callbacks are registered that will write to the console indicating main
eSPI events and when a virtual wire is received from the master.
```

dbkinder commented on samples/drivers/espi/README.rst:
```diff
@@ -0,0 +1,45 @@
+Building and Running
+********************
+
+The sample can be built and executed on boards supporting eSPI
```
```suggestion
The sample can be built and executed on boards supporting eSPI.
```

dbkinder commented on samples/drivers/espi/README.rst:
```diff
@@ -0,0 +1,45 @@
+********************
+
+The sample can be built and executed on boards supporting eSPI
+If any pins are required for minimum eSPI handshake these can be configured
```
```suggestion
Any pins required for minimum eSPI handshake should be configured
```

dbkinder commented on samples/drivers/espi/README.rst:
```diff
@@ -0,0 +1,45 @@
+        2nd phase completed
+        3rd phase completed
+
+note:: The values shown above might differ.
```
```suggestion
note:: The values shown above might differ on your board.
```

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+	enum espi_io_mode io_caps;
+	/** Supported channels */
+	enum espi_channel channel_caps;
+	/** Maximum supported frequency in MHz*/
```
Change end of line to MHz */

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+ *
+ * If this eSPI controller is acting as slave, the values set here
+ * will be discovered as part through the GET_CONFIGURATION command
+ * issued by the eSPI master during initialization
```
add period

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+ *
+ * If this eSPI controller is acting as master, the values set here
+ * will be used by eSPI master to determine minimum common capabilities with
+ * eSPI slave then send via SET_CONFIGURATION command
```
add period

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+/**
+ * @brief Query to see if it a channel is ready.
+ *
+ * This routine allows to check if logical channel is ready before use
```
add periods

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+ * @brief Query to see if it a channel is ready.
+ *
+ * This routine allows to check if logical channel is ready before use
+ * Note that queries for channels not supported will always return false
```
add periods

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+/**
+ * @brief Sends memory, I/O or message read request over eSPI.
+ *
+ * This routines provides a generic interface to send a read request packet
```
add periods

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+ *
+ * @retval 0 If successful.
+ * @retval -ENOTSUP if eSPI controller doesn't support raw packets and instead
+ *         low memory transactions are handled by controller hardware directly
```
add period

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+ * @retval 0 If successful.
+ * @retval -ENOTSUP if eSPI controller doesn't support raw packets and instead
+ *         low memory transactions are handled by controller hardware directly
+ * @retval -EIO General input / output error, failed to send over the bus
```
add periods

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+/**
+ * @brief Sends memory, I/O or message write request over eSPI.
+ *
+ * This routines provides a generic interface to send a write request packet
```
add periods

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+	if (!api->write_request) {
+		return -ENOTSUP;
+	}
+	return api->write_request(dev, req);
```
add space between the if and the return

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+ * @brief Sends system/platform signal as a virtual wire packet.
+ *
+ * This routines provides a generic interface to send a virtual wire packet
+ * from slave to master
```
period

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+ * @param level The level of signal requested LOW or HIGH.
+ *
+ * @retval 0 If successful.
+ * @retval -EIO General input / output error, failed to send over the bus
```
period

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+ * @brief Retrieves level status for a signal encapsulated in a virtual wire.
+ *
+ * This routines provides a generic interface to request a virtual wire packet
+ * from eSPI master and retrieve the signal level
```
period

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+	if (!api->send_oob) {
+		return -ENOTSUP;
+	}
+	return api->send_oob(dev, pckt);
```
add space between if and return

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+	if (!api->receive_oob) {
+		return -ENOTSUP;
+	}
+	return api->receive_oob(dev, pckt);
```
add space between if and return

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+	if (!api->flash_read) {
+		return -ENOTSUP;
+	}
+	return api->flash_read(dev, pckt);
```
add space between and return

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+	if (!api->flash_write) {
+		return -ENOTSUP;
+	}
+	return api->flash_write(dev, pckt);
```
add space between if and return in all the occurrences

franciscomunoz commented on CODEOWNERS:
```diff
@@ -113,6 +113,7 @@
 /drivers/counter/                         @nordic-krch
 /drivers/display/                         @vanwinkeljan
 /drivers/dma/*sam0*                       @Sizurka
+/drivers/espi/                            @albertofloyd @franciscomunoz
```
remember we are only owners of mchp_espi_xec. We already have an expression for all the drivers   /drivers/*/*xec* . So this line should be removed

franciscomunoz commented on CODEOWNERS:
```diff
@@ -232,6 +233,7 @@
 /include/init.h                           @andrewboie @andyross
 /include/irq.h                            @andrewboie @andyross
 /include/irq_offload.h                    @andrewboie @andyross
+/include/espi.h                           @albertofloyd @franciscomunoz
```
add scott

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+ * Note: To help setting it, see espi_init_callback() below
+ */
+struct espi_callback {
+	/** This is meant to be used in the driver only */
```
is up to you if you want to put period to single line comments line this one. But be consistent

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+	espi_api_flash_erase flash_erase;
+	espi_api_manage_callback manage_callback;
+};
+/**
```
leave a space between }; and the comment below

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+ * @endcond
+ */
+
+
```
Remove one line

franciscomunoz commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+ *
+ * +--------+   +---------+     +------+          +---------+   +---------+
+ * |  eSPI  |   |  eSPI   |     | eSPI |          |  eSPI   |   |  eSPI   |
+ * |  sla^e |   | dri^er  |     |  bus |          |  dri^er |   |  host   |
```
correct |  sla^e |   | dri^er  |     |  bus |          |  dri^er |

tbursztyka commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+ */
+typedef int (*espi_api_config)(struct device *dev, struct espi_cfg *cfg);
+typedef bool (*espi_api_get_channel_status)(struct device *dev,
+						enum espi_channel ch);
```
indentation should be below last '(' + 1 space. Apply that everywhere

tbursztyka commented on include/espi.h:
```diff
@@ -0,0 +1,829 @@
+	if (!api->read_request) {
+		return -ENOTSUP;
+	}
+	return api->read_request(dev, req);
```
have an empty line after a '}' unless it's another '}' etc... Apply that everywhere.

albertofloyd replied:
@tbursztyka  I re-checked and space guideline after if/else blocks is followed everywhere

andrewboie replied:
@tbursztyka can you please limit awarding patches -1s to *egregious* style issues, and incorrect code?
We're not going to make progress if we're going to sit here arguing about what blank line goes where

tbursztyka replied:
I thought these comments were already applied long ago. The ones that were not applied were enum names and DTS stuff (which is a valid issue to -1)

tbursztyka reviewed (changes requested):
The API, driver and samples are fine by me. I don't have anything to comment there.

So now, just apply the style comment I had (on espi.h and in the driver) and that will be good to go.

dbkinder reviewed (approved):
doc changes LGTM, thanks!

andrewboie commented:
@franciscomunoz @tbursztyka can you revisit?

franciscomunoz commented on samples/drivers/espi/src/main.c:
```diff
@@ -0,0 +1,226 @@
+	espi_add_callback(espi_dev, &vw_cb);
+	espi_add_callback(espi_dev, &p80_cb);
+	printk("complete\n");
+	return ret;
```
add space between printk and return

albertofloyd commented:
> The API, driver and samples are fine by me. I don't have anything to comment there.
> 
> So now, just apply the style comment I had (on espi.h and in the driver) and that will be good to go.

All style issues are now addressed

franciscomunoz commented on samples/drivers/espi/src/main.c:
```diff
@@ -0,0 +1,230 @@
+		printk("SLP_S5 Timeout!");
+		return ret;
+	}
+	ret = wait_for_vwire(espi_dev, ESPI_VWIRE_SIGNAL_SLP_S4,
```
leave space after }

franciscomunoz commented on samples/drivers/espi/src/main.c:
```diff
@@ -0,0 +1,230 @@
+		printk("SLP_S3 Timeout!");
+		return ret;
+	}
+	printk("\t2nd phase completed\n");
```
leave space after }

franciscomunoz commented on samples/drivers/espi/src/main.c:
```diff
@@ -0,0 +1,230 @@
+	}
+}
+
+/* Indicate to eSPI master only single line, 20MHz frequency and
```
Indicate to eSPI master the following configuration : 
Single Line, 20MHz frequency and logical channel 0.

franciscomunoz commented on samples/drivers/espi/README.rst:
```diff
@@ -0,0 +1,44 @@
+********************
+
+The sample can be built and executed on boards supporting eSPI.
+Any pins required for minimum eSPI handshake should be configured
```
add period

franciscomunoz commented on drivers/espi/espi_utils.h:
```diff
@@ -0,0 +1,67 @@
+
+
+/**
+ * @brief Generic function to insert or remove a callback from a callback list
```
Add periods in the @brief and @param sentences describing the APIs

tbursztyka commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,831 @@
+}
+
+static bool espi_xec_channel_ready(struct device *dev,
+					enum espi_channel ch)
```
there is still indentation issue: that parameter should be below the 's' of struct. (so below last '(' + 1 space).ù

Your editor should be doing that for you if you configure it to follow linux code style.

albertofloyd replied:
Are you suggesting to mix tabs and spaces here? I haven't seen that anywhere in the rest of Zephyr code.

albertofloyd replied:
Addressed anyways to follow linux kernel closely

tbursztyka replied:
tabs and spaces afterwards are hopefully allowed and everywhere in zephyr.

spaces + tabs or tabs + spaces + tabs are not.

tbursztyka commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,831 @@
+		(xec_id >= ESPI_MSVW_IDX_MAX)) {
+		return -EINVAL;
+	}
+	if (signal_info.dir == ESPI_MASTER_TO_SLAVE) {
```
always add an empty line after '}' unless it's another '}' etc... Apply that everywhere.

albertofloyd replied:
Addressed

tbursztyka reviewed (changes requested):
still style issues. some comments were not followed

albertofloyd commented:
> still style issues. some comments were not followed

Added parameter alignment for functions in test app.

andrewboie commented:
@tbursztyka please re-visit your -1

tbursztyka commented:
@andrewboie still looks like some of my comments on the espi driver haven't been followed. 

@albertofloyd check the espi driver, the enum name (not the enum attributes) should be in small letters. And also one comment on IRQ not coming from DTS. (Btw, do you see the comments? it could be github which is messing up things here)

andrewboie commented:
this PR needs a rebase

albertofloyd commented on drivers/espi/espi_mchp_xec.c:
```diff
@@ -0,0 +1,822 @@
+	ESPI_EIO_BAR_REGS->EC_BAR_P80CAP_0 = 0x00810001;
+}
+
+static void setup_espi_io_config(u16_t host_address)
```
Fixed

albertofloyd committed:
boards : arm : Add ESPI support for MEC15xx EVB

Expose eSPI block with interrupts enabled for channel 0 & 1
eSPI handshake has been tested using espi driver sample app

Signed-off-by: Jose Alberto Meza <jose.a.meza.arellano@intel.com>
